[clang] Function type attribute to prevent CFI instrumentation

PiJoules opened the pull request:
This introduces the attribute discussed in
https://discourse.llvm.org/t/rfc-function-type-attribute-to-prevent-cfi-instrumentation/85458.

The proposed name has been changed from `no_cfi` to `cfi_unchecked_callee` to help differentiate from `no_sanitize("cfi")` more easily. The proposed attribute has the following semantics:

1. Indirect calls to a function type with this attribute will not be instrumented with CFI. That is, the indirect call will not be checked. Note that this only changes the behavior for indirect calls on pointers to function types having this attribute. It does not prevent all indirect function calls for a given type from being checked.
2. All direct references to a function whose type has this attribute will always reference the true function definition rather than an entry in the CFI jump table.
3. When a pointer to a function with this attribute is implicitly cast to a pointer to a function without this attribute, the compiler will give a warning saying this attribute is discarded. This warning can be silenced with an explicit C-style cast or C++ static_cast.

PiJoules committed:
[clang] Function type attribute to prevent CFI instrumentation

This introduces the attribute discussed in
https://discourse.llvm.org/t/rfc-function-type-attribute-to-prevent-cfi-instrumentation/85458.

The proposed name has been changed from `no_cfi` to
`cfi_unchecked_callee` to help differentiate from `no_sanitize("cfi")`
more easily. The proposed attribute has the following semantics:

1. Indirect calls to a function type with this attribute will not be
   instrumented with CFI. That is, the indirect call will not be
   checked. Note that this only changes the behavior for indirect calls
   on pointers to function types having this attribute. It does not
   prevent all indirect function calls for a given type from being checked.
2. All direct references to a function whose type has this attribute will
   always reference the true function definition rather than an entry
   in the CFI jump table.
3. When a pointer to a function with this attribute is implicitly cast
   to a pointer to a function without this attribute, the compiler
   will give a warning saying this attribute is discarded. This warning
   can be silenced with an explicit C-style cast or C++ static_cast.

PiJoules commented:
@erichkeane We wanted C++ support so I tried adding exhaustive tests but I'm not sure if I caught everything. Would you be able to give a pass and see if I missed anything and what else should be tested? Also I'm having trouble coming up with a "struct redefinition" test. I think I'm misinterpreting N3037, but that looks like it has to do with type compatibility between different TUs rather than types within the same TU?

erichkeane reviewed (commented):
I'm concerned about the approach of trying to leave this as `AttributedType`, as that is going to result in some particularly sharp edges to this.  Consider:

```
using WithAttr = __attribute__((cfi_unchecked_callee)) BaseType;
using WithoutAttr = __attribute__((cfi_unchecked_callee)) BaseType;


using MyType = WithAttr;
using MyType = WithoutAttr; // still valid, since the attribute hasn't changed the type.
```

The diagnostic-inconsistency here mixed with the strange-never-before-an-odr-like-violation odr-like-violation is going to be extra weird.

This, IMO, needs its own AST type, with its own proper conversion rules.

@aaronballman for his opinion if it differs.

erichkeane commented:
> @erichkeane We wanted C++ support so I tried adding exhaustive tests but I'm not sure if I caught everything. Would you be able to give a pass and see if I missed anything and what else should be tested? Also I'm having trouble coming up with a "struct redefinition" test. I think I'm misinterpreting N3037, but that looks like it has to do with type compatibility between different TUs rather than types within the same TU?

N3037 allows redefinition of the same struct within the same TU.  We only recently ahve an implementation of it (not sure if it has been merged yet).

I did a quick look, I didn't see anything about variable-type-aliases, static member functions, or explicit-this-type member functions, operators, etc.

efriedma-quic commented:
I also think this needs to be a real type.  (Probably you can just add a bit to FunctionTypeExtraBitfields.)

ilovepi reviewed (commented):
For the new test files, is it possible to precommit them? I think it will be easier to review the delta.

PiJoules commented:
I updated to this with a new AST node but would like some guidance in implementation details. One of the shortcomings of making this a (sugared) node is when passing this type around through templates, the node doesn't persist since only canonical types are substituted through templates. This could lead to something like:

```
void func() __attribute__((cfi_unchecked_callee)) {}
template <typename T>
struct S {
  S(T *ptr) {}
};
S s(&func);
```

Giving a warning since clang deduces `T *ptr` as the canonical `void (*)()`. Other attributes like `noreturn` work around this by being part of the function prototype via `ExtInfo` or `FunctionTypeExtraBitfields` as @efriedma-quic suggested, but using the bitfields alone wouldn't allow warning on casts since those attrributes get freely added/dropped without warning. I think perhaps the best way to approach this would be instead treating this as an extra bitfield rather than having a node and for every implicit conversion which would add or drop attributes, specifically check if this attribute is dropped. I'd like some opinions on this before moving forward though just to make sure I don't do unnecessary work. I imagine we wouldn't want a mix of both a node and extra bitfield since we'd have two representations of the same thing. Another idea is perhaps making this not a sugared type, but I don't know if there would be ramifications to that.

efriedma-quic commented:
I'm confident making this part of the FunctionType is the right approach.

We don't implicitly drop the bits from the FunctionType bitfields, for bits that are part of the canonical type.  Any cast is explicitly represented in the AST, and we have warnings for mismatches (see checkCastFunctionType()).  The warnings are just off by default because weird casts are common in practice.

PiJoules commented:
Ok rather than making the attr its own AST node, I found it much simpler to make it part of the function type. Rather than adding it to `FunctionTypeExtraBitfields` I instead added it to `ExtInfo`. I did since it looks like if I wanted to add a bit to `FunctionTypeExtraBitfields`, then I'd also need to set a bit in `ExtProtoInfo` to indicate the extra bitfields should be allocated, but that feels like I'd be adding an extra redundant bit. Rather than adding it in `ExtProtoInfo` though, I moved it to  `ExtInfo` since that would also cover the no-prototype functions and there were some unused bits in that same struct.

> I did a quick look, I didn't see anything about variable-type-aliases, static member functions, or explicit-this-type member functions, operators, etc.

I also added more Frontend test cases to cover these. Lemme know if there's other coverage I'm missing.

There's some stuff in this patch that is also somewhat independent of the attribute (like the `IsFunctionConversion` changes), so if this patch looks ok at a high level, I can start breaking it up into more independent patches.

PiJoules commented:
ping for initial reviews

PiJoules commented:
> For the new test files, is it possible to precommit them? I think it will be easier to review the delta.

I'll precommit once folks think this is ok at a high level. Wouldn't want to potentially rewrite tests if the implementation happens to change a lot.

ilovepi commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -3006,6 +3006,10 @@ static LValue EmitFunctionDeclLValue(CodeGenFunction &CGF, const Expr *E,
                                      GlobalDecl GD) {
   const FunctionDecl *FD = cast<FunctionDecl>(GD.getDecl());
   llvm::Constant *V = CGF.CGM.getFunctionPointer(GD);
+  if (E->getType()->isCFIUncheckedCalleeFunctionType()) {
+    if (auto *GV = dyn_cast<llvm::GlobalValue>(V))
+      V = llvm::NoCFIValue::get(GV);
+  }
   CharUnits Alignment = CGF.getContext().getDeclAlign(FD);
   return CGF.MakeAddrLValue(V, E->getType(), Alignment,
```
since it get's used in the return, you could cache `E->getType()`.

```suggestion
  auto ETy = E->getType();
  if (ETy->isCFIUncheckedCalleeFunctionType()) {
    if (auto *GV = dyn_cast<llvm::GlobalValue>(V))
      V = llvm::NoCFIValue::get(GV);
  }
  CharUnits Alignment = CGF.getContext().getDeclAlign(FD);
  return CGF.MakeAddrLValue(V, ETy, Alignment,
```
Feel free to use a better name and explicit type.

PiJoules replied:
Done

ilovepi commented on clang/lib/CodeGen/ItaniumCXXABI.cpp:
```diff
@@ -693,6 +693,18 @@ CGCallee ItaniumCXXABI::EmitLoadOfMemberFunctionPointer(
+
+  if (ShouldEmitCFICheck) {
+    if (const auto *BinOp = dyn_cast<BinaryOperator>(E)) {
+      if (BinOp->isPtrMemOp()) {
+        if (BinOp->getRHS()
```
Maybe combine these?

PiJoules replied:
Done

PiJoules commented on clang/include/clang/AST/Type.h:
```diff
@@ -1988,7 +1988,7 @@ class alignas(TypeAlignment) Type : public ExtQualsTypeCommonBase {
     /// regparm and the calling convention.
     LLVM_PREFERRED_TYPE(CallingConv)
-    unsigned ExtInfo : 14;
+    unsigned ExtInfo : 15;
```
Hmm. I suppose I could move it to `ExtProtoInfo` in `FunctionPrototype` and forgo attaching this to non function prototypes. I would hope adding it there instead doesn't add extra overhead since it aleady has a couple of members already and we wouldn't need to bring in another trailing object like what's needed for `FunctionTypeExtraBitfields`.

PiJoules replied:
Updated to use `ExtProtoInfo` instead so `ExtInfo` remains unchanged.

PiJoules commented:
ping for reviews

erichkeane reviewed (approved):
I only have minor suggestions, else this looks good to me.  Please give Aaron a chance to take another  look as well.

PiJoules commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -3027,9 +3027,13 @@ static LValue EmitFunctionDeclLValue(CodeGenFunction &CGF, const Expr *E,
                                      GlobalDecl GD) {
   const FunctionDecl *FD = cast<FunctionDecl>(GD.getDecl());
   llvm::Constant *V = CGF.CGM.getFunctionPointer(GD);
+  auto ETy = E->getType();
```
Done

PiJoules commented on clang/include/clang/AST/Type.h:
```diff
@@ -8253,6 +8275,19 @@ inline bool Type::isObjectPointerType() const {
+  return false;
+}
+
+inline bool Type::isPointerToCFIUncheckedCalleeFunctionType() const {
```
Not sure how I should rename it though. Perhaps something like `hasPointeeToToCFIUncheckedCalleeFunctionType`?

PiJoules replied:
Renamed and updated the body

AaronBallman replied:
Naming this sort of stuff is always extremely painful. See https://github.com/llvm/llvm-project/pull/122938 for other fun crimes.

ilovepi commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12711,6 +12711,11 @@ def warn_noderef_on_non_pointer_or_array : Warning<
 
+def warn_cast_discards_cfi_unchecked_callee
+    : Warning<"implicit conversion from %0 to %1 discards "
+              "`cfi_unchecked_callee` attribute">,
```
These look like backticks ... should they be single quotes? That's the pattern I see above for `noderef`. Was wondering if that's why the docs build broke in the premerge check.
```suggestion
              "'cfi_unchecked_callee' attribute">,
```

PiJoules replied:
Updated. Didn't know ticks vs single quotes mattered to clang docs.

AaronBallman reviewed (approved):
LGTM

brunodf-snps commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -9283,8 +9283,14 @@ static AssignConvertType checkPointerTypesForAssignment(Sema &S,
   }
-  if (!S.getLangOpts().CPlusPlus && S.IsFunctionConversion(ltrans, rtrans))
-    return AssignConvertType::IncompatibleFunctionPointer;
+  bool DiscardingCFIUncheckedCallee, AddingCFIUncheckedCallee;
+  if (!S.getLangOpts().CPlusPlus &&
+      S.IsFunctionConversion(ltrans, rtrans, &DiscardingCFIUncheckedCallee,
+                             &AddingCFIUncheckedCallee)) {
+    // Allow conversions between CFIUncheckedCallee-ness.
+    if (!DiscardingCFIUncheckedCallee && !AddingCFIUncheckedCallee)
+      return AssignConvertType::IncompatibleFunctionPointer;
+  }
```
Hi, I'm late here because we only spotted this while integrating clang-21 in our downstream tree.

I think there is a logic error here. If a function conversion is _also_ discarding or adding CFIUncheckedCallee status, it is suddenly no longer an incompatible function pointer? That does not make sense to me, and it stops diagnosing conversions that we should be diagnosing (and normally are diagnosing), see this example: https://godbolt.org/z/bnYsE63Gc

Granted, the original code has flawed logic, which I reported already back in 2019: https://bugs.llvm.org/show_bug.cgi?id=42120
In short, the code seems to use the _negative_ of the `IsFunctionConversion` condition in the _inverse_ direction of what you would expect. My bug report observes that this appears to work but breaks in more complex cases. I suggested a patch at the time, but this did not get picked up. I would consider submitting a PR, but I would like some clang maintainers to agree with my report and patch (or suggest some other fix), before doing so.

brunodf-snps replied:
Follow up in #160474 and #160477, please review.
